# Project 1 Group 10

#### Members: Harshit Aggarwal, Ranchen Huang, Wenqing Qin, Yinghang Qiu, Arun Rajan


##### check working directory

In [2]:
!pwd

/home/ubuntu/notebooks/Assignments/Assignment 2


## Problem 1

### Part A: Characters in The Hound of the Baskervilles (15 points)

Use the text available at https://s3.amazonaws.com/dmfa-2020/project-1/hound.txt for this part.

How many times are each of the following characters mentioned by name in the text of The Hound of the Baskervilles?

Holmes, Watson, Barrymore, Mortimer

Hint: Be careful where the names matches both the exact name and a plural relating to a family name. For example, Watsons vs. Watson.  What do you observe?

##### Upload text file using !wget command

In [3]:
!wget https://s3.amazonaws.com/dmfa-2020/project-1/hound.txt

--2021-09-27 01:48:01--  https://s3.amazonaws.com/dmfa-2020/project-1/hound.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.160.61
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.160.61|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 345659 (338K) [text/plain]
Saving to: ‘hound.txt’

hound.txt           100%[===================>] 337.56K  --.-KB/s    in 0.01s   

2021-09-27 01:48:02 (28.7 MB/s) - ‘hound.txt’ saved [345659/345659]



##### Check head of the file text

#### Holmes

##### Search for how many times of 'Holmes' appears

In [4]:
!grep Holmes hound.txt | wc

    190    2203   12785


* Holmes is mentioned in the text 190 times.

In [5]:
!grep -w Holmes hound.txt | wc

    190    2203   12785


#### Search for how many times of 'Watson'appears

In [6]:
!grep Watson hound.txt | wc

    113    1343    7580


* Watson is mentioned in the text 113 times.

In [40]:
!grep -w Watson hound.txt | wc

    113    1343    7580


#### Search for how many times of 'Barrymore'appears

In [7]:
!grep Barrymore hound.txt | wc

     78     875    5106


* Barrymore is mentioned in the text 72 times.

In [10]:
!grep -w Barrymore hound.txt | wc 

     72     806    4691


#### Search for how many times of 'Mortimer'appears

In [43]:
!grep Mortimer hound.txt | wc

     90    1036    6143


* Mortimer is mentioned in the text 89 times.

In [45]:
!grep -w Mortimer hound.txt | wc

     89    1023    6075


### Part B: Characters in Hamlet (15 points)

Use the text available at https://s3.amazonaws.com/dmfa-2020/project-1/hamlet.txt for this part.

How many times do each of the following characters have speaking lines in Hamlet? Keep in mind that this is the text of a play.

Hamlet, Polonius, Ophelia, and Horatio

A speaking line is usually start with an Upper-case name with a trailing  “.”
For example, 
HAMLET. Not so, my lord, I am too much i’ the Sun

##### Upload text file using !wget command

In [22]:
!wget https://s3.amazonaws.com/dmfa-2020/project-1/hamlet.txt

--2021-09-27 02:46:23--  https://s3.amazonaws.com/dmfa-2020/project-1/hamlet.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.217.42.30
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.217.42.30|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 204885 (200K) [text/plain]
Saving to: ‘hamlet.txt’

hamlet.txt          100%[===================>] 200.08K  --.-KB/s    in 0.009s  

2021-09-27 02:46:23 (21.7 MB/s) - ‘hamlet.txt’ saved [204885/204885]



In [23]:
# Check head of the file text
!head hamlet.txt


Project Gutenberg EBook of Hamlet, by William Shakespeare

This eBook is for the use of anyone anywhere in the United States and
most other parts of the world at no cost and with almost no
restrictions whatsoever.  You may copy it, give it away or re-use it
under the terms of the Project Gutenberg License included with this
eBook or online at www.gutenberg.org.  If you are not located in the
United States, you’ll have to check the laws of the country where you
are located before using this ebook.


#### Hamlet


In [24]:
!grep -F "HAMLET. " hamlet.txt | wc

    358    3537   19856


* Hamlet has 358 speaking lines.


#### Polonius

In [25]:
!grep -F "POLONIUS. " hamlet.txt | wc

     86     823    4837


* Polonius has 86 speaking lines.


#### Ophelia


In [26]:
!grep -F "OPHELIA. " hamlet.txt | wc

     58     542    3135


* Ophelia has 58 speaking lines.

#### Horatio

In [27]:
!grep -F "HORATIO. " hamlet.txt | wc

    107     892    5096


* Horatio has 107 speaking lines.

## Problem 2

### Part A

1. Unzip the data and combine the two inflating CSV files using csvstack. Name your combined file “trips.csv”. Compare the total number of lines in each file and in the new combined file. Comment on your findings.

In [ ]:
# Get the data
!wget https://s3.amazonaws.com/dmfa-2020/project-1/2018-capitalbikeshare-tripdata.zip

In [ ]:
# Get the numbers of files
!unzip -l 2018-capitalbikeshare-tripdata.zip

In [ ]:
# Unzip the data
!unzip 2018-capitalbikeshare-tripdata.zip

In [ ]:
# Using csvstack combine the data
!csvstack 201807-capitalbikeshare-tripdata.csv 201808-capitalbikeshare-tripdata.csv > trips.csv

In [ ]:
# Get the total number of lines in the individual files and the final combined file
!wc -l 201807-capitalbikeshare-tripdata.csv
!wc -l 201808-capitalbikeshare-tripdata.csv
!wc -l trips.csv

The 2 files individually have 404762 and 403867 lines of data whereas the combined file has 808628 lines.
There is a discrepancy of 1 line, which can be attributed the the header column.

2. List the labels for the heading line.  

In [15]:
!csvcut -n trips.csv

  1: Duration
  2: Start date
  3: End date
  4: Start station number
  5: Start station
  6: End station number
  7: End station
  8: Bike number
  9: Member type


3. Which 12 Capital Bikeshare stations were the most popular departing stations in July and August 2018 in terms of number of rides? provide the full name of the station, not just the station number.

In [4]:
!csvcut -c4,5 trips.csv | sort | uniq -c | sort -r | head -12

  15474 31258,Lincoln Memorial
  13082 31623,Columbus Circle / Union Station
  11999 31247,Jefferson Dr & 14th St SW
  11671 31289,Henry Bacon Dr & Lincoln Memorial Circle NW
  11650 31288,4th St & Madison Dr NW
  10357 31248,Smithsonian-National Mall / Jefferson Dr & 12th St SW
   9554 31249,Jefferson Memorial
   8620 31290,17th St & Independence Ave SW
   8583 31200,Massachusetts Ave & Dupont Circle NW
   8569 31201,15th & P St NW
   7711 31603,1st & M St NE
   7578 31321,15th St & Constitution Ave NW


4. Which 12 stations were the most popular destination stations in July and August 2018 in terms of number of rides?  provide the full name of the station, not just the station number. 

In [ ]:
!csvcut -c6,7 trips.csv | sort | uniq -c | sort -r | head -12

5. Which 12 station-pairs (Departing-Destination) are most popular in July and August 2018 in terms of number of rides?  provide the full name of the station, not just the station number.

In [ ]:
!csvcut -c5,7 trips.csv | sort | uniq -c | sort -r | head -12

6. Comments on your finding for Q3, Q4 and Q5.

Even though Lincoln Memorial and Columbus Circle/Union Station are the top departing and destination stations it is not the most popular route to travel with them not being in the top 12. Lincoln Memorial to Jefferson Memorial seems like a popular router given that they appear in both the lists. Lincoln Memorial back to Lincoln Memorial is the 5th most populus route.

### Part B

1. For the most popular departure station, which 10 bikes were used most in trips departing from there? provide the full name of the station, not just the station number.

In [ ]:
# From Part A question 3, get the most popular departure station: 31258, Lincoln Memorial
# get the info with 4,5,8 column
!csvgrep -c4 -r "31258" trips.csv | csvcut -c4,5,8 | sort | uniq -c | sort -r | head -10

2. Which 10 bikes were used most in trips ending at the most popular destination station? provide the full name of the station, not just the station number.

In [ ]:
!csvgrep -c6 -r "31258" trips.csv | csvcut -c6,7,8 | sort | uniq -c | sort -r | head -10

## Problem 3

#### Question 1 :
Upload the two files into your root directory, combine the first file (header) with the unzipped version of the second file (data) and name the combined file as “candidate.csv” (7 points)
Hint: you may need to reformat your second file from “|” delimited to csv before you combine it with the first file. You can use csvformat command. 

The following were performed
* Get the cm_header_file.csv from the weblink
* Get the cm22.zip file from the weblink
* Unzip the cm22.zip file using unzip command

In [ ]:
!wget https://www.fec.gov/files/bulk-downloads/data_dictionaries/cm_header_file.csv
!wget https://www.fec.gov/files/bulk-downloads/2022/cm22.zip
!unzip cm22.zip

We then convert the txt file to cm.csv in csv format. Then, the 15 columns in the file is listed out. The 15th column remains unnamed or empty.

In [ ]:
!csvformat -d "|" cm.txt > cm.csv
!csvcut -n cm.csv

The names of the 15 columns in the file cm_header_file.csv is listed out.

In [ ]:
!csvcut -n cm_header_file.csv

The data from the cm.csv is stacked below the cm_header_file.csv, and combined into a file named candidate.csv

In [ ]:
!csvstack cm_header_file.csv cm.csv > candidate.csv

#### Question 2. 
Provide a  list of the label heading for the combined file. Display the first 5 columns and 20 rows of the combined file 

Display the first 5 columns and 20 rows of the combined file using the csvcut to select columns 1 to 5, and then display the output using csvlook. 

In [ ]:
!csvcut -n candidate.csv 

The first 5 columns and 20 rows of the combined file is obtained using the head command, which is piped to csvcut for columns 1 to 5 only, and the results obtained using csvlook is listed below.

In [6]:
!head -20 candidate.csv | csvcut -c 1-5 | csvlook

| CMTE_ID   | CMTE_NM                                                                                         | TRES_NM                     | CMTE_ST1                          | CMTE_ST2        |
| --------- | ----------------------------------------------------------------------------------------------- | --------------------------- | --------------------------------- | --------------- |
| C00000422 | AMERICAN MEDICAL ASSOCIATION POLITICAL ACTION COMMITTEE                                         | WALKER, KEVIN MR.           | 25 MASSACHUSETTS AVE, NW          | SUITE 600       |
| C00000489 | D R I V E POLITICAL FUND CHAPTER 886                                                            | JERRY SIMS JR               | 3528 W RENO                       |                 |
| C00000547 | KANSAS MEDICAL SOCIETY POLITICAL ACTION COMMITTEE                                               | JERRY SLAUGHTER             | 623 SW 10TH AVE                   |                 |
| C00000729 | AMERIC

#### Question 3

Provide a list of all Republican committees with an office in NY. 
The list should contain the Committee name, Committee city or town, and Connected organization's name. Sort the finding by Comittee name.


We use csvgrep to identify the 'REP' in column 11 and 'NY' in column 7 of candidate.csv file. 
Then, using csvcut, the columns 2, 6 and 14 are selected.
csvlook is used to display the result of Republican in NY

In [ ]:
!csvgrep -c11 -r "(?i)REP" candidate.csv | csvgrep -c7 -r "(?i)NY" candidate.csv | csvcut -c 2,6,14 | csvsort -c 1 | csvlook 

#### Question 4
Provide the names of the committees that appear more than two times in this dataset? If you find any investigate whether these records are duplicates and suggest an action (e.g. remove) (10 points). 

We use the csvcut to select Column 2 of candidate.csv which has the Committee name, and then use sort, uniq and sort -rn to i
grep and identify the multiple entries greater than 2.

In [8]:
!csvcut -c2 candidate.csv | sort | uniq -c | sort -rn | grep "2" -v | grep "1" -v | wc -l

3


In [9]:
!csvcut -c2 candidate.csv | sort | uniq -c | sort -rn | grep "2" -v | grep "1" -v

      4 COLLINS FOR CONGRESS
      3 SCHNEIDER FOR CONGRESS
      3 HOLD THEM ACCOUNTABLE


Now, we find that there are three Committees which appear more than two times. 

In [10]:
!grep "COLLINS FOR CONGRESS" candidate.csv | sort -k 2,2

C00705236,JOE COLLINS FOR CONGRESS,"MARSTON, CHRIS",5777 CENTURY BLVD,STE 1110-143,LOS ANGELES,CA,90045,P,H,REP,Q,,NONE,H0CA43108
C00781104,KINA COLLINS FOR CONGRESS,"COLLINS, KINA",901 LAKE ST,UNIT 4093,OAK PARK,IL,60301,P,H,DEM,Q,,KINA FOR CONGRESS,H0IL07167
C00785006,JAY COLLINS FOR CONGRESS,"KILGORE, PAUL",7853 GUNN HWY,BOX 238,TAMPA,FL,33626,P,H,REP,Q,,NONE,H2FL14251
C00502039,COLLINS FOR CONGRESS,"DANIEL, CARROLL BRIAN",PO BOX 1295,,GAINESVILLE,GA,30503,P,H,REP,T,,,H2GA09150
C00521641,COLLINS FOR CONGRESS,CYNTHIA M GIFFORD,176 SYCAMORE LANE,,WAKEFIELD,RI,02879,P,H,IND,A,,,H2RI02143
C00544684,COLLINS FOR CONGRESS,"BROWN, MEGAN",PO BOX 6491,,ATHENS,GA,30604,P,H,REP,Q,,NONE,H4GA10071
C00611137,COLLINS FOR CONGRESS,CATHERINE ANN PAGE,3045 DARIEN PARK DRIVE,,ROSWELL,GA,30076,P,H,REP,Q,,,H4GA11103


In [11]:
!grep 'SCHNEIDER FOR CONGRESS\|HOLD THEM ACCOUNTABLE' candidate.csv | sort -k 2,2 -k 3,3 -s

C00672956,SCHNEIDER FOR CONGRESS,"CRATE, BRADLEY T",PO BOX 317,,BEVERLY,MA,01915,P,H,REP,Q,,NONE,H8MA06133
C00495952,SCHNEIDER FOR CONGRESS,"KARTON, DEBORAH",PO BOX 1318,,DEERFIELD,IL,60015,P,H,DEM,Q,,NONE,H2IL10068
C00374751,SCHNEIDER FOR CONGRESS,"SCHNEIDER, JAN MS",227 SEAGULL LANE,,SARASOTA,FL,34236,A,H,DEM,Q,,,H2FL13139
C00690263,HOLD THEM ACCOUNTABLE,"VOGEL, TARYN",PO BOX 2505,,MT PLEASANT,SC,29465,U,V,,Q,,,
C00771923,HOLD THEM ACCOUNTABLE,"VOGEL, TARYN",PO BOX 2505,,MT PLEASANT,SC,29465,U,O,,T,,,
C00764175,HOLD THEM ACCOUNTABLE,"WINTER, DAVID MR",15 PARK STREET APT 206,,ESSEX JCT.,VT,05452,U,O,,T,,,


### Conclusion for Question 4: 

* Although there seem to be as many as 10 names of committees that appear more than two times, it is understood that in 8 cases, the Candidate names and addresses are unique, and therefore can be ignored.

* In the case of Vogel, Taryn, the name of Committee remains same for two separate records. However, since the Committee ID is unique, the records from this database can only indicate duplicity. It will have to be verified and set right based on actual documented paperwork. 

* After checking if deemed duplicate we can go ahead and remove this record from our data